fluent/context.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import absolute_import, unicode_literals
+
+import babel
+import babel.numbers
+import babel.plural
+import six
+
+from .builtins import BUILTINS
+from .resolver import resolve
+from .syntax import FluentParser
+from .syntax.ast import Message, Term
+from .utils import cachedproperty
+
+
+class MessageContext(object):
+    """
+    Message contexts are single-language stores of translations.  They are
+    responsible for parsing translation resources in the Fluent syntax and can
+    format translation units (entities) to strings.
+
+    Always use `MessageContext.format` to retrieve translation units from
+    a context.  Translations can contain references to other entities or
+    external arguments, conditional logic in form of select expressions, traits
+    which describe their grammatical features, and can use Fluent builtins.
+    See the documentation of the Fluent syntax for more information.
+    """
+
+    def __init__(self, locales, functions=None, use_isolating=True):
+        self.locales = locales
+        _functions = BUILTINS.copy()
```

Comment:
I copy them because the next line mutates the dictionary to add in the custom ones i.e. I want a dictionary that combines the builtins and the custom ones, without mutating the builtins (or the passed in custom dictionary, for that matter, as general good practice).

`dict.copy()` does a shallow copy, so the values (which are functions) are not cloned in any way, plus the keys are immutable (strings) so they are not cloned either. So you just get a new independent dict object, but no copying of the actual keys or values.